### PR TITLE
Fix RUF010 auto-fix with parenthesis

### LINKED ISSUE #4520: [Autofix error] RUF010 autofix fails when the expression is in parentheses, e.g. `f"{(str(a))}"`
RUF010 creates an invalid autofix when the expression is already in parentheses, e.g.
```python
f"{(str(a))}"
```
is replaced by 
```python
f"{(a!s)}"
```
which is invalid syntax.

Command:

```
ruff --no-cache --select RUF010 --fix code.py
```

ruff version fd16d658e9407ce76bc38cd8f5112dd272fc5be0

### PR DESCRIPTION
**Pull request by JonathanPlasse**
- Close #4520
- From #4423
  > It looks like we now generate one diagnostic for the whole f-string (and that the diagnostic range is that of the entire f-string).
  > 
  > I think it'd be preferable to retain one diagnostic per formatted value, and limit the diagnostic to the formatted value. It's fine if the autofix contents include the whole string, even for each formatted value, but they should be separate diagnostics. (E.g., we could iterate over `formatted_values` and create one diagnostic per index, with a fix specific to that index.)

### CODE REVIEW

File: crates/ruff/src/cst/matchers.rs [hunk 1]
```diff
@@ -111,3 +112,33 @@ pub(crate) fn match_simple_string<'a, 'b>(
...
+    }
+}
+
+pub(crate) fn match_name<'a, 'b>(expression: &'a mut Expression<'b>) -> Result<&'a mut Name<'b>> {
+    if let Expression::Name(name) = expression {
+        Ok(name)
+    } else {
+        bail!("Expected Expression::Name")
+    }
+}
```
Comment (MichaReiser) [inline crates/ruff/src/cst/matchers.rs:144]: Can we avoid adding these new methods and instead pass a `ast::FormattedString` node as argument OR, if the rule should accept multiple nodes, define a new union that only contains the nodes relevant for that rule?

File: crates/ruff/src/rules/ruff/rules/explicit_f_string_type_conversion.rs [hunk 2]
```diff
@@ -39,59 +47,124 @@ impl AlwaysAutofixableViolation for ExplicitFStringTypeConversion {
...
+    }
+
+    let mut state = CodegenState {
+        default_newline: &stylist.line_ending(),
+        default_indent: stylist.indentation(),
+        ..CodegenState::default()
+    };
+    expression.codegen(&mut state);
+
+    Ok(Fix::automatic(Edit::range_replacement(
```
Comment (MichaReiser) [inline crates/ruff/src/rules/ruff/rules/explicit_f_string_type_conversion.rs:89]: Is there a chance that this refactor changes the semantic of the code?

Comment (JonathanPlasse) [reply crates/ruff/src/rules/ruff/rules/explicit_f_string_type_conversion.rs:89]: It should not.
https://peps.python.org/pep-3101/#explicit-conversion-flag

File: crates/ruff/src/rules/ruff/rules/explicit_f_string_type_conversion.rs [hunk 3]
```diff
@@ -39,59 +47,124 @@ impl AlwaysAutofixableViolation for ExplicitFStringTypeConversion {
...
 
-    if checker.patch(diagnostic.kind.rule()) {
-        let arg_range = args[0].range();
-        let remove_call = Edit::deletion(formatted_value_range.start(), arg_range.start());
-        let add_conversion = Edit::replacement(
-            format!("!{conversion}"),
-            arg_range.end(),
-            formatted_value_range.end(),
+    if checker.patch(diagnostics[0].kind.rule()) {
+        let fix = fix_explicit_f_string_type_conversion(
```
Comment (charliermarsh) [inline crates/ruff/src/rules/ruff/rules/explicit_f_string_type_conversion.rs:153]: I think we should generate a separate fix for each diagnostic. As-is, the diagnostic only highlights one violation (correct), but the fix shows modifying the whole string.

**Pull request merged**